services/gundeck/src/Gundeck/ThreadBudget.hs

```diff
@@ -0,0 +1,212 @@
+-- http://hackage.haskell.org/package/rate-limit also looks related.
+--
+-- USE CASE: keep a lid of stalled native push notification threads.  if SNS is up, there
+-- will be many short-running executions of the action.  when SNS is down, the threads will
+-- accumulate in memory and choke the gundeck instances.  so we want to stop spawning more
+-- threads (and discard or queue native push notifications) before we run out of memory (which
+-- could cause system outages).
+--
+-- http-client connection pools should handle this naturally and without doing anything, but
+-- instead connection pools grow infinitely until system resources (file handles, memory) are
+-- exhausted.  See https://github.com/snoyberg/http-client/issues/307#issuecomment-343829351
+module Gundeck.ThreadBudget
+  ( ThreadBudgetState
+  , mkThreadBudgetState
+  , runWithBudget
+  , watchThreadBudgetState
+
+  -- * for testing
+  , threadBudgetLimits
+  , runningThreads
+  , cancelAllThreads
+  ) where
+
+import Imports
+
+import Control.Exception (ErrorCall(ErrorCall))
+import Control.Exception.Safe (catchAny)
+import Control.Lens
+import Control.Monad.Catch (MonadCatch, throwM)
+import Data.Metrics (Metrics)
+import Data.Metrics.Middleware (gaugeSet, path)
+import Data.SizedHashMap (SizedHashMap)
+import Data.Time
+import Data.UUID (UUID, toText)
+import Data.UUID.V4 (nextRandom)
+import Gundeck.Options
+import UnliftIO.Async
+import UnliftIO.Exception (finally)
+
+import qualified Data.Set as Set
+import qualified Data.SizedHashMap as SHM
+import qualified System.Logger.Class as LC
+
+
+data ThreadBudgetState = ThreadBudgetState
+  { _threadBudgetLimits  :: MaxConcurrentNativePushes
+  , _threadBudgetRunning :: IORef BudgetMap
+  } deriving (Generic)
+
+-- | Store all handles for cleanup in 'watchThreadBudgetState'.
+type BudgetMap = SizedHashMap UUID (Maybe (Async ()))
+
+-- generated with @makeLenses@, but we only want one of the two lenses that generates.
+threadBudgetLimits ::
+  Lens' ThreadBudgetState MaxConcurrentNativePushes
+threadBudgetLimits f_a1A3A (ThreadBudgetState x1_a1A3B x2_a1A3C)
+  = (fmap (\ y1_a1A3D -> (ThreadBudgetState y1_a1A3D) x2_a1A3C))
+      (f_a1A3A x1_a1A3B)
+{-# INLINE threadBudgetLimits #-}
+
+-- | Instead of taking the size of the SizedHashMap (O(1)), this counts all the threads that
+-- are successfully running (dropping the ones that are just about to try to grab a token,
+-- taking O(n)).
+runningThreads :: ThreadBudgetState -> IO Int
+runningThreads (ThreadBudgetState _ running)
+  = length . filter isJust . SHM.elems <$> readIORef running
+
+cancelAllThreads :: ThreadBudgetState -> IO ()
+cancelAllThreads (ThreadBudgetState _ ref) = readIORef ref
+  >>= mapM_ cancel . catMaybes . SHM.elems
+
+showDebugHandles :: BudgetMap -> String
+showDebugHandles = show . SHM.size
+
+
+mkThreadBudgetState :: HasCallStack => MaxConcurrentNativePushes -> IO ThreadBudgetState
+mkThreadBudgetState limits = if limits ^. mcnpHard < limits ^. mcnpSoft
```

Comment:
Hum, we currently allow `0` or `-1` meaning infinite.  This raises the point though that we should check that integers are positive, so I'll do at least that.

Additionally, I could allow both limits to be optional individually, and change the semantics so that soft limit always logs, and hard limit always blocks.  I would use `null` instead fo `0`, though.  I don't think this change is worth the effort, so please let me know if you disagree, then i'll do it anyway (not *that* much trouble).

Reply:
Making `hard` (or both) optional sounds good!